### PR TITLE
Fix nn.init.trunc_normal_ call on torch.float16 data

### PR DESCRIPTION
**Pull request by fxmarty**
Following https://github.com/huggingface/transformers/pull/20803 that gave the idea, but was still buggy for some cases, for example:

```python
from transformers import ViTForMaskedImageModeling
import torch

model = ViTForMaskedImageModeling.from_pretrained('hf-internal-testing/tiny-random-vit', torch_dtype=torch.float16).to("cuda")
```

still raising `RuntimeError: "erfinv_vml_cpu" not implemented for 'Half'`.

Let me know if you would like me to add tests.

### CODE REVIEW

File: src/transformers/models/vit/modeling_vit.py
Comment (ydshieh) [inline src/transformers/models/vit/modeling_vit.py:462]: Just a few lines above this change, we have already the same change shown below, so the changes in this PR is the way to go! Thank you.

```python
        if isinstance(module, (nn.Linear, nn.Conv2d)):
            # Upcast the input in `fp32` and cast it back to desired `dtype` to avoid
            # `trunc_normal_cpu` not implemented in `half` issues
            module.weight.data = nn.init.trunc_normal_(
                module.weight.data.to(torch.float32), mean=0.0, std=self.config.initializer_range
            ).to(module.weight.dtype)
```

**Review by ydshieh — APPROVED** [review]

**Review by sgugger — APPROVED** [review]
Thanks for the fixes!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your great work on this! 💯

**Comment by fxmarty** [comment]
@sgugger @younesbelkada Feel free to merge, it appears I have no merge rights even after approval on transformers.

**Pull request merged**
